Add diff task to find differences between two sets of UUIDs

i-like-robots opened the pull request:
This new task will take two files (output by the `uuids` task), find additions/removals and then advise on what action to take.

Partly ported from the [autosync in ES interface][1]. One step towards to fulfilling #3.

Usage:

```sh
$ n-es-tools uuids eu
> output file uuids-eu.txt
$ n-es-tools uuids us
> output file uuids-us.txt
$ n-es-tools diff uuids-eu.txt uuids-us.txt
> 755b39d8-f27f-11e0-931e-00144feab49a should be ingested
> 755b39d8-f27f-11e0-931e-00144feab49a should be deleted
```

Underneath it uses the unix diffutils. Ideally I would have liked to make this multiplatform but I couldn't find a way (in JS) of finding a reliable diff for such large files.

[1]: https://github.com/Financial-Times/next-es-interface/blob/master/server/lib/sync.js

leggsimon commented:
RE
```
> 755b39d8-f27f-11e0-931e-00144feab49a should be ingested
> 755b39d8-f27f-11e0-931e-00144feab49a should be deleted
```
Should this say in which cluster it should be ingested or deleted

i-like-robots commented:
Yeah that would be more useful. I'll add it.

I did some more testing (pasting here so I don't get to work and have to write it again...) and I came up with this crappy script:

```js
const fs = require('fs')

const memory = () => (process.memoryUsage().rss / 1048576).toFixed(2)

console.log(`Memory: ${memory()}mb`)

const a = fs.readFileSync('uuids-eu.txt').toString().split('\n')
const b = fs.readFileSync('uuids-us.txt').toString().split('\n')

const start = Date.now()

const uniqueA = new Set(a)
const uniqueB = new Set()

console.log(`${a.length} items loaded into set`)
console.log(`Memory: ${memory()}mb`)

let i = 0
let len = b.length

while (i < len) {
  if (uniqueA.has(b[i])) {
    uniqueA.delete(b[i])
  } else {
    uniqueB.add(b[i])
  }

  i++
}

const end = Date.now()

console.log(`Time taken: ${(end - start) / 1000}s`)
console.log(`Final memory: ${memory()}mb`)

console.log('Differences A:', Array.from(uniqueA))
console.log('Differences B:', Array.from(uniqueB))
```

And it works surprisingly well!:

```
Memory: 19.68mb
678337 items loaded into set
Memory: 244.93mb
Time taken: 1.153s
Final memory: 265.92mb
Differences A: [ '3ded0dd2-a2ef-11de-ba74-00144feabdc0' ]
Differences B: [ 'f3d36225-8445-3220-aa20-06ba712e8297' ]
```

The memory usage means dropping onto a Heroku dyno or Lambda might not always work but locally that's not a lot to worry about.

i-like-robots commented:
Less memory:

```js
const fs = require('fs')
const readline = require('readline')

const memory = () => (process.memoryUsage().rss / 1048576).toFixed(2)

const loadFile = (filename, cb) => {
  return new Promise((resolve, reject) => {
    const input = fs.createReadStream(filename)
    const reader = readline.createInterface({ input })

    input.on('error', reject)

    reader.on('line', cb)
    reader.on('close', resolve)
  })
}

console.log(`Memory: ${memory()}mb`)

const start = Date.now()

const uniqueA = new Set()
const uniqueB = new Set()

const handleA = (line) => {
  uniqueA.add(line)
}

const handleB = (line) => {
  if (uniqueA.has(line)) {
    uniqueA.delete(line)
  } else {
    uniqueB.add(line)
  }
}

Promise.resolve()
  .then(() => loadFile('uuids-eu.txt', handleA))
  .then(() => {
    console.log(`${uniqueA.size} items loaded into set`)
    console.log(`Memory: ${memory()}mb`)
  })
  .then(() => loadFile('uuids-us.txt', handleB))
  .then(() => {
    const end = Date.now()

    console.log(`Time taken: ${(end - start) / 1000}s`)
    console.log(`Final memory: ${memory()}mb`)

    console.log('Differences A:', Array.from(uniqueA))
    console.log('Differences B:', Array.from(uniqueB))
  })
```

Result:

```
Memory: 19.71mb
678336 items loaded into set
Memory: 135.95mb
Time taken: 1.918s
Final memory: 167.06mb
Differences A: [ '3ded0dd2-a2ef-11de-ba74-00144feabdc0' ]
Differences B: [ 'f3d36225-8445-3220-aa20-06ba712e8297' ]
```

i-like-robots commented:
Closing in favour of #6
